File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    var answer = [];
    let totalPlays =new Map();
    let setPlayList = new Map();
    for(let i=0;i<genres.length;i++){
        if(totalPlays.get(genres[i])===undefined) totalPlays.set(genres[i],plays[i]);
        else totalPlays.set(genres[i],totalPlays.get(genres[i])+plays[i]);
        
        if(setPlayList.get(genres[i])===undefined) setPlayList.set(genres[i],[[i,plays[i]]]);
        else setPlayList.set(genres[i],[...setPlayList.get(genres[i]),[i,plays[i]]]);
    }
    let totalPlaysToArr = [...totalPlays];
    totalPlaysToArr.sort((a,b)=>b[1]-a[1]);
    setPlayList.forEach((m)=>{
        m.sort((a,b)=>{
            if(a[1]===b[1]) return a[0]-b[0];
            else return b[1]-a[1];
        })
    })
    for(let i=0;i<totalPlaysToArr.length;i++){
        const playList = setPlayList.get(totalPlaysToArr[i][0]);
        const ans = playList.slice(0,2);
        console.log(ans);
        ans.map((e)=>{
            answer.push(e[0]);
        })
        // answer.push(playList[0][0]);
        // if(playList.length >1)answer.push(playList[1][0]);
    }
    return answer;
}
/*
sol
1. 장르별로 총 재생 횟수를 저장 후 저장 횟수별로 내림차순 정렬 ex) classic: 2000
2. 해당 장르별 고유 번호와 재생 횟수를 저장
3. 
*/

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
#include <algorithm>
using namespace std;
bool compare(pair<string, int> a, pair<string, int> b){
    return a.second > b.second;
}
bool compareList(pair<int, int> a,pair<int, int> b){
    if(a.second == b.second) return a.first < b.first;
    else return a.second > b.second;
}
map<string,int> m; // 아이템별 총 플레이수
map<string,vector<pair<int,int>>> playList;
vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    for(int i = 0;i<genres.size();i++){
        m[genres[i]]+=plays[i];
        
        playList[genres[i]].push_back(make_pair(i,plays[i]));
        
    }
    vector<pair<string, int>> v(m.begin(), m.end());
    sort(v.begin(),v.end(),compare);
    for(auto genre : playList) {
        vector<pair<int, int>> sortedList = genre.second; // 복사 생성
        sort(sortedList.begin(), sortedList.end(), compareList); // 정렬
        playList[genre.first] = sortedList; // 정렬된 리스트를 다시 할당
    }
    
    for(auto g : playList){
        for(auto gg : g.second){
            cout << gg.first << " ";
        }
    }
    for(auto g : v){
        answer.push_back(playList[g.first][0].first);
        if(playList[g.first].size() > 1) answer.push_back(playList[g.first][1].first);
    }
    
    return answer;
}
